File: collectors/youtube.py
"""YouTube trending collector using Data API v3."""
from typing import List
from googleapiclient.discovery import build
from .base import BaseCollector, NewsItem


class YouTubeCollector(BaseCollector):
    """Fetch trending tech videos."""
    
    def __init__(self, config: dict):
        super().__init__(config)
        self.youtube = None
        if self.enabled and config.get("api_key"):
            try:
                self.youtube = build("youtube", "v3", developerKey=config["api_key"], static_discovery=False)
            except Exception as e:
                print(f"[YouTube] Init failed: {e}")
    
    def is_available(self) -> bool:
        return self.enabled and self.youtube is not None
    
    def collect(self) -> List[NewsItem]:
        if not self.is_available():
            return []
        
        items = []
        region = self.config.get("region_code", "US")
        category = self.config.get("category_id", "28")  # Science & Tech
        max_results = self.config.get("max_results", 10)
        
        try:
            resp = self.youtube.videos().list(
                part="snippet,statistics",
                chart="mostPopular",
                regionCode=region,
                videoCategoryId=category,
                maxResults=max_results
            ).execute()
            
            for item in resp.get("items", []):
                try:
                    snippet = item["snippet"]
                    stats = item.get("statistics", {})
                    video_id = item["id"]
                    
                    title = snippet.get("title", "")
                    url = f"https://youtube.com/watch?v={video_id}"
                    
                    views = int(stats.get("viewCount", 0))
                    likes = int(stats.get("likeCount", 0))
                    hot_score = views * 0.001 + likes * 0.1
                    
                    items.append(NewsItem(
                        title=title,
                        url=url,
                        source="youtube",
                        source_name=snippet.get("channelTitle", "YouTube"),
                        hot_score=round(hot_score, 1),
                        category="technology",
                        language=snippet.get("defaultLanguage", "en"),
                        raw_data=stats
                    ))
                except Exception:
                    continue
                    
        except Exception as e:
            print(f"[YouTube] Error: {e}")
        
        return items

File: lib/collectors/youtube.ts
import type { Collector } from "./base";
import { createNewsItem } from "./base";

const API_BASE = "https://youtube.googleapis.com/youtube/v3";

export class YouTubeCollector implements Collector {
  name = "youtube";
  private config: Record<string, unknown>;
  private apiKey: string | null = null;

  constructor(config: Record<string, unknown>) {
    this.config = config;
    this.apiKey = (config.api_key as string) ?? null;
  }

  async collect() {
    if (!this.apiKey) return [];

    const items = [];
    const region = (this.config.region_code as string) ?? "US";
    const category = (this.config.category_id as string) ?? "28"; // Science & Tech
    const maxResults = (this.config.max_results as number) ?? 10;

    try {
      const params = new URLSearchParams({
        part: "snippet,statistics",
        chart: "mostPopular",
        regionCode: region,
        videoCategoryId: category,
        maxResults: String(maxResults),
        key: this.apiKey,
      });

      const resp = await fetch(`${API_BASE}/videos?${params}`, {
        signal: AbortSignal.timeout(15_000),
      });

      if (!resp.ok) {
        console.error(`[YouTube] HTTP ${resp.status}: ${await resp.text()}`);
        return [];
      }

      const data = (await resp.json()) as {
        items?: {
          id: string;
          snippet: {
            title: string;
            channelTitle: string;
            defaultLanguage?: string;
          };
          statistics?: {
            viewCount?: string;
            likeCount?: string;
            commentCount?: string;
          };
        }[];
      };

      for (const item of data.items ?? []) {
        try {
          const snippet = item.snippet;
          const stats = item.statistics ?? {};
          const videoId = item.id;

          const title = snippet.title ?? "";
          const url = `https://youtube.com/watch?v=${videoId}`;
          const views = parseInt(stats.viewCount ?? "0", 10);
          const likes = parseInt(stats.likeCount ?? "0", 10);
          const hotScore = views * 0.001 + likes * 0.1;

          items.push(
            createNewsItem(title, url, "youtube", snippet.channelTitle ?? "YouTube", {
              hot_score: Math.round(hotScore * 10) / 10,
              category: "technology",
              language: snippet.defaultLanguage ?? "en",
              raw_data: stats as Record<string, unknown>,
            })
          );
        } catch {
          // per-item error: skip
        }
      }
    } catch (err) {
      console.error(`[YouTube] Error: ${err}`);
    }

    return items;
  }
}
